Reworks barbering

Superlagg opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->
[REWORK][FEATURE]
## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Scissors and razors can be toggled between barber and murder mode by using it in-hand. They default to murder mode.

In barber mode, you can style anyone's hair (even yourself!) anywhere you want, not just in a barber chair.

However, barbering is not an automatic success, and the conditions in which you barber influence whether you cut their hair the way you intended, or just mangle their hair and shank up their head.

Conditions include:
- If the subject is dead or the barber is a robot (automatic success)
- If you and/or your subject is in the barber shop (less of a bonus if you're somehow not in the same room as your subject)
- If you spawned as a barber and/or you're wearing the barber outfit
- If you or your subject is jittery (no penalty if you both are jittery though)
- If you're clumsy

Barbering checks if the subject is a mutant, and has a response for all the current mutant races.

The razor now properly picks from a list of facial hair styles and not the whole list.

Barbering uses an action bar!

Hairdye has been tweaked. Now, you can use the bottle in-hand to switch between target hair layers. It can also be set to dye all three layers or the eyes.

If you're a barber, you get three uses out of a full bottle of dye, unless you have the bottle set to dye all the hair. Otherwise, you get one use out of it before needing to refill it.

I'm not sure if you could before, but you can hit people with an empty dye bottle. It doesn't do anything fancy, but its possible I guess.

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

I got distracted trying to make barbering work for my limbs'n'tails PR, and then this happened cus it looked neat.

## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```
(u)Superlagg:
(*)Barbering can be done anywhere now! Though it should really be done by a barber in the barber shop.
(+)Barber tools can be switched between barber and murder mode by clicking them in-hand. They start set to murder, so be careful!
(+)Dye bottles can target hair layers, and can be used three times if you're a barber.
```

pali6 commented:
Just some random thoughts, not a full review: 
- This also makes it so you switch surgical scissors between barber mode an murder mode which is not very intuitive.
- The action bar for cutting hair seems to be all kinds of fucked. It doesn't get interrupted by the target moving out of range but it does get interrupted by you just clicking on the floor to turn which is a bit annoying.
- Trying to give haircuts to non-human mobs (e.g. cyborgs) runtimes (line 645).
- Cutting hair of someone buckled into the barber chair doesn't do anything and doesn't even provide you with a message of your success / failure.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -123,200 +185,137 @@
-		return
-	if(src.empty)
+
+/obj/item/proc/dye_hair(mob/living/carbon/human/M as mob, mob/user as mob, obj/item/dye_bottle/bottle as obj)
```
Why is this definted on /obj/item ????

Superlagg replied:
Not sure why the dye bottle's defined there, I cant think of any other items that'd be suitable.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -123,200 +185,137 @@
+		var/fuckedyourshitup = 0
+
+		if(user.bioHolder.HasEffect("clumsy" && prob(40)))
+			var/degreetowhichwefuckedtheirshitup = 0
```
Funny name haha but also it makes this not easy to read.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -123,200 +185,137 @@
+		var/result_msg1 = "[user] dyes [M]'s hair."
+		var/result_msg2 = "<span class='notice'>You dye [M]'s hair.</span>"
+		var/result_msg3 = "<span class='notice'>[user] dyes your hair.</span>"
+		var/is_barber = user.mind.assigned_role == "Barber" ? 1 : 0
```
` ? 1 : 0` is redundant.

Superlagg replied:
Just the first conditional returns what I want? Neat!

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+
+
+// Barber stuff
+/obj/item/proc/knife_fluff(mob/user)
```
Again why is this defined on `/obj/item`?

Superlagg replied:
That's so it could theoretically be used with other items, I do intend to allow things like surgical scissors and butcher knives to be used for barbering.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+		else
+			boutput(user, "<span class='alert'>You wield the [src] with [pick(cool_grip_adj)] [pick(cool_grip1)] [pick(cool_grip2a)][pick(cool_grip2b)] [pick("style", "grip")] that you [pick(cool_grip3)]! It makes it just about impossible to cut hair or do surgery!</span>")
+
+/obj/item/proc/scissor_action(mob/living/carbon/human/M as mob, mob/living/carbon/human/user as mob)
```
Ditto.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+		return 0
+
+
+/obj/item/proc/razor_action(mob/living/carbon/human/M as mob, mob/living/carbon/human/user as mob)
```
Just all of this.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+		return 0
+
+
+/obj/item/proc/do_shave(mob/living/carbon/human/M as mob, mob/living/carbon/human/user as mob)
```
...

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+	actions.start(new/datum/action/bar/shave(M, user, degree_of_success, new_style), user)
+	return 1
+
+/obj/item/proc/do_haircut(mob/living/carbon/human/M as mob, mob/living/carbon/human/user as mob)
```
aaaaaaa

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+
+/proc/get_barbery_conditions(mob/living/carbon/human/M as mob, mob/living/carbon/human/user as mob)
+	var/barbery_conditions = 0
+	// let's see how ideal the haircutting conditions are
```
Mixing addition, subtration and multiplication like this is not very intuitive and I pity the person that's gonna explain it on the wiki. I recommend axing `*=`s.

Superlagg replied:
True, and it does lead to some weird numbers. I suppose I could just make the multipliers just flat addition and subtraction.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -437,3 +436,555 @@
+				barbery_conditions *= 0.5
+
+	barbery_conditions = clamp(barbery_conditions, 1, 100)
+	return (barbery_conditions)
```
`(` `)` is redundant here.

Superlagg replied:
Forgot I left it like that, since that proc used to be part of the haircut proc.

pali6 commented on code/obj/item.dm:
```diff
@@ -105,6 +107,16 @@
+	var/list/wheredWeSeeIt = list("saw the clown do", "saw the captain do", "saw a large dog do", "saw the head of security do",\
+														"saw someone in a red spacesuit do", "saw a floating saw do", "saw on TV",\
+														"saw one of the diner dudes do", "saw just about every assistant do")
+
```
Not more stuff defined on /obj/item please. 😭

Superlagg replied:
I didn't want to define the lists in the proc that uses them since they are fairly long lists and the proc can be done repeatedly, I wasn't sure if the game needed to rebuild the lists every time the proc fired.

I'll move it to the proc.

Superlagg commented:
> Just some random thoughts, not a full review:
> 
> * This also makes it so you switch surgical scissors between barber mode an murder mode which is not very intuitive.

Forgot that surgical scissors are a child type of scissors, seeing as the razor has no children. But in a way its what I wanted to do with them anyway. I could add in some more checks to make the resulting text more fitting to the tool though, and add in a description and sprite transformation to make it more obvious what mode its in.

> * The action bar for cutting hair seems to be all kinds of fucked. It doesn't get interrupted by the target moving out of range but it does get interrupted by you just clicking on the floor to turn which is a bit annoying.

Figures something'd be wrong with it, I *just* figured out how to make actionbars work. I'll fix it!

> * Trying to give haircuts to non-human mobs (e.g. cyborgs) runtimes (line 645).

ffs

> * Cutting hair of someone buckled into the barber chair doesn't do anything and doesn't even provide you with a message of your success / failure.

_ffs_

Superlagg commented:
Found out why it kept failing, sometimes the barbery conditions would end up returning a -1, which bypassed the switch entirely. Fixed that!

Scissors and razors default to having force_use_as_tool (previously do_barbery) disabled, but added a check to see if the user is on help intent and if their target is either in the barber shop or in a barber chair to prevent accidentally attacking the person. 

Fixed the action bar to interrupt if the distance between you and your target becomes more than 1. Still interrupts if you click on other things while its running, but so does most everything else, quit fidgeting and cut their hair!

Setting force_use_as_tool makes the tool face a different direction. Also scissors and razors are 4-directional and can face different directions when thrown. 

Anything with force_use_as_tool resets it to attack-mode when thrown or dropped. Attack-mode should work just like how the tool already works, aside from the barbershop checks.

Made the razor and scissor checks owned by razors and scissors. Didn't do that with the actual haircut procs, since they could be useful for other things, like wirecutters or scalpels. Not that they are, yet.

Barbery conditions now only use addition and subtraction, and runs the prob check itself. Also, failure just picks a number between 0-2, with 2 being more likely.

Cutting a cyborg's (or AI's) hair, instead of throwing runtimes, returns a message.

Also medical cyborgs can cut hair, apparently.

Superlagg commented:
Moved all the procs and such to a component and now call them with signals. Works about the same, but without nearly as much shit defined in /obj/item.

Gerhazo commented on code/obj/item.dm:
```diff
@@ -30,10 +30,12 @@
 	pressure_resistance = 50
 	var/obj/item/master = null
 	var/amount = 1
+	var/uses_left
```
Only the dye bottle seems to make use of this. Is defining at item level necessary?

Superlagg replied:
Forgot about that part. Fixed it!

pali6 commented:
Instead of the weirdly specific signal TOOLONLY I'd have the component register the proc to dropped and attack_self signals. (Attack_self might not exist yet.)
Throw_begin shouldn't be necessary I think (in a sane world dropped gets called first).

pali6 commented:
Triggering the BARBER signal from attack manually sorta misses the point of signals and components. The idea behind components is that I should be able to add the barbering component to a sledgehammer and bam now the sledgehammer can be used to shave people. Consider refactoring so the signal is triggered in mob's attackby.

pali6 commented:
Barber Vs shave can just be subtypes of the main barbering component.

pali6 commented:
Actually it should likely just be on some ATTACK signal and not a new signal at all. However, that might prove troublesome due to the difficulty of preventing other attack actions happening and just general ordering of things. I'll try to clean up some signal stuff to allow for better hooks at various points in attacking things with other things.

pali6 commented on code/atom.dm:
```diff
@@ -704,6 +704,8 @@
 
 //mbc : sorry, i added a 'is_special' arg to this proc to avoid race conditions.
 /atom/proc/attackby(obj/item/W as obj, mob/user as mob, params, is_special = 0)
+	if((SEND_SIGNAL(src, COMSIG_ATTACKBY, W, user, params, is_special)) & ATTACK_PRE_DONT_ATTACK)
```
Wait hold on, this probably shouldn't be here? At least not in this PR. Or if in this PR it probably shouldn't use the flag labeled as ATTACK_PRE...

Superlagg replied:
Forgot I left that in there. It was when I was trying to see i-- wait shit I forgot if I actually used that or not. One of the components does listen for COMSIG_ATTACKBY. Uh. I removed the che-- wait, no I was right the first time, I don't use that anymore. Removed!

Superlagg commented:
Changed the toolmode component to listen to attack_self and dropped signals to either toggle the mode or disable it. No need for extra signals in the item!

Also attack_self sends a signal about mid-way through it.

And toolmode more properly prevents accidental murder. I mean with barbertools, attacking someone on help intent pretty much always tries to cut their hair (cus you can do it anywhere), but toolmode prevents you from hurting people on every intent.

pali6 commented on code/obj/ToSplit/barber_shop.dm:
```diff
@@ -64,9 +88,16 @@
 		BLOCK_KNIFE
 
+	attack(mob/M as mob, mob/user as mob)
+		if (scalpel_surgery(M, user))
+			return 1
```
Is not calling parent intended?

Superlagg replied:
Mustve missed that, made razors not work. Now its there and it does!

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+	// this proc is supposed to make certain tools less accidentally deadly for inexperienced players to use
+	// when force_use_as_tool is set, all intents will try to do their tool-thing, and if it can't, return a message saying they're using it wrong
+	// if not set, help intent will still attempt tool, but you'll shank them if it doesn't work out
+/datum/component/toggle_tool_use/proc/on_drop(var/obj/item/thing)
```
This should take an additional `mob/user` argument.

pali6 replied:
```suggestion
/datum/component/toggle_tool_use/proc/on_drop(var/obj/item/thing, mob/user)
```

Superlagg replied:
Even though it doesn't use it? Aight.

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+	// when force_use_as_tool is set, all intents will try to do their tool-thing, and if it can't, return a message saying they're using it wrong
+	// if not set, help intent will still attempt tool, but you'll shank them if it doesn't work out
+/datum/component/toggle_tool_use/proc/on_drop(var/obj/item/thing)
+	toggle_force_use_as_tool(thing)
```
Changing behaviour based on whether you supply the `user` argument is really unintuitive. Especially considering that this could just be replaced with
```suggestion
	thing.force_use_as_tool = 0
	thing.dir = SOUTH

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+	toggle_force_use_as_tool(thing, user)
+
+/datum/component/toggle_tool_use/proc/toggle_force_use_as_tool(var/obj/item/thing, mob/user)
+	if(!user)
+		thing.force_use_as_tool = 0
+		thing.dir = SOUTH
+		return
```
As I said, I'd move this to on_drop and remove it from here.

Superlagg replied:
Done!

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+		thing.dir = SOUTH
+		return
+	thing.force_use_as_tool = !thing.force_use_as_tool
+	if (thing.force_use_as_tool)
+		thing.dir = WEST
+	else
+		thing.dir = SOUTH
```
This dir changeing definitely deserves a comment explaining it because it's not clear at all. I assume you are toggling different "icon states" based on the tool mode? Doing it through dir seems really janky but I guess it can stay even if I don't like it.

An alternative cleaner way of doing this would be for the component to take arguments in `Initialize` that say what the icon state is supposed to be in tool mode and what it's supposed to be in non-tool mode and then storing and switching between those.

Superlagg replied:
I used dir as a way to both change what it looks like in hand, and to let the tools look different when thrown. After all, changing the way you hold a pair of scissors doesn't really change how the thing looks, just what direction it's pointing.

I suppose it would be kind of weird for tools that'd need to change icon_states when toggled. But, those don't exist _yet_, and using the component to change icon_state would need anything that does change icon_state to use a specific naming format (scissors and scissors-toolmode), and/or a defined icon_state_toolmode var. And might cause problems if those arent defined.

I could define and explain the dirs in init, though.

pali6 replied:
Probably just add a comment for now.

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+	if(ishuman(M) && ((H.head && H.head.c_flags & COVERSEYES) || (H.wear_mask && H.wear_mask.c_flags & COVERSEYES) || (H.glasses && H.glasses.c_flags & COVERSEYES)))
+		// you can't stab someone in the eyes wearing a mask!
+		boutput(user, "<span class='notice'>You're going to need to remove that mask/helmet/glasses first.</span>")
+		. = BARBERY_RESOLVABLE
```
When it comes to this and similar lines below - is BARBERY_RESOLVABLE actually checked anywhere? It seems like this proc should only return `ATTACK_PRE_DONT_ATTACK` or `0` probably?

Superlagg replied:
BARBERY_RESOLVABLE means that barbery failed, but gave a useful message about why. Should be checked lower down, first if (.) is not 0, then if (.) != BARBERY_RESOLVABLE, throw a generic message and return ATTACK_PRE_DONT_ATTACK.

pali6 replied:
But you also do `return BARBERY_SUCCESSFUL` later.

pali6 replied:
I'd just use a new var for these instead of `.`. When I see `. = stuff` I assume you are planning to return `stuff`.

Superlagg replied:
Aight, replaced them with a var.

Also it turned out that I didnt actually need most of those defines. Aside from the initial pre-checks, they werent even used! All that got cut and replaced with a var.

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+		if (new_style)
+			if(M.bioHolder.mobAppearance.customization_first == "Balding" && new_style != "None")
+				boutput(user, "<span class='alert'>Not enough hair!</span>")
+				return ATTACK_PRE_DONT_ATTACK
```
Note that this is in a spawn so it won't actually return the value. I recommend replacing just with 
```suggestion
				return
```

pali6 commented on code/datums/components/barber.dm:
```diff
@@ -0,0 +1,605 @@
+				return ATTACK_PRE_DONT_ATTACK
+		else
+			boutput(user, "Never mind.")
+			return ATTACK_PRE_DONT_ATTACK
```
```suggestion
			return
```

pali6 reviewed (approved):
I think this is ok, hopefully you tested it more than I did.
